Shoko.Server/Plugin/Loader.cs

```diff
@@ -18,36 +20,42 @@ public class Loader : ISettingsProvider
         private readonly IList<Type> _pluginTypes = new List<Type>();
         private static readonly ILogger Logger = LogManager.GetCurrentClassLogger();
         
-        internal void Load(IServiceCollection serviceCollection)
+        internal void Load(IServiceCollection serviceCollection, IConfiguration config)
         {
             var assemblies = new List<Assembly>();
             var assembly = Assembly.GetExecutingAssembly();
             var uri = new UriBuilder(assembly.GetName().CodeBase);
             var dirname = Path.GetDirectoryName(Uri.UnescapeDataString(uri.Path));
             // if (dirname == null) return;
             assemblies.Add(Assembly.GetCallingAssembly()); //add this to dynamically load as well.
-            
+            var pluginConfig = config.GetSection("Plugins");
+            var pluginSettings = pluginConfig.Get<PluginSettings>();
+
             //Load plugins from the user config dir too.
             var userPluginDir = Path.Combine(ServerSettings.ApplicationPath, "plugins");
-            var userPlugins = Directory.Exists(userPluginDir) ? Directory.GetFiles(userPluginDir) : new string[0];
+            var userPlugins = Directory.Exists(userPluginDir) ? Directory.GetFiles(userPluginDir, "*.dll") : new string[0];
             
             foreach (var dll in Directory.GetFiles(dirname, "plugins/*.dll", SearchOption.AllDirectories).Concat(userPlugins))
             {
                 try
                 {
                     string name = Path.GetFileNameWithoutExtension(dll);
-                    if (ServerSettings.Instance.Plugins.EnabledPlugins.ContainsKey(name) &&
-                        !ServerSettings.Instance.Plugins.EnabledPlugins[name])
+                    if (pluginSettings.EnabledPlugins.ContainsKey(name) && pluginSettings.EnabledPlugins[name] == false)
                     {
                         Logger.Info($"Found {name}, but it is disabled in the Server Settings. Skipping it.");
                         continue;
                     }
                     Logger.Debug($"Trying to load {dll}");
                     assemblies.Add(Assembly.LoadFrom(dll));
                     // TryAdd, because if it made it this far, then it's missing or true.
-                    ServerSettings.Instance.Plugins.EnabledPlugins.TryAdd(name, true);
-                    if (!ServerSettings.Instance.Plugins.Priority.Contains(name))
-                        ServerSettings.Instance.Plugins.Priority.Add(name);
+                    if (!pluginSettings.EnabledPlugins.ContainsKey(name))
+                        pluginConfig.GetSection("EnabledPlugins")[name] = "true";
+                    if (!pluginSettings.Priority.Contains(name))
+                    {
+                        pluginSettings.Priority.Add(name);
+                        for (int i = 0; i < pluginSettings.Priority.Count; i++)
+                            pluginConfig.GetSection("Priority")[i.ToString()] = pluginSettings.Priority[i];
```

Comment:
This won't work for IConfiguration, as it is readonly

Reply:
I've confirmed it's working on my machine. 🤷

Reply:
https://github.com/aspnet/Configuration/issues/385

Reply:
It may work on your machine, but it won't persist